resources/top_level_overview.markdown

```diff
@@ -0,0 +1,283 @@
+Thus for each input split, a map task is spawned.  The number of maps is
+typically driven by the number of HDFS blocks in the input file. This block size
+is configurable, and therefore allows for a workaround when wishing to adjust the
+number of map tasks.
+
+There does exists a `mapred.map.tasks` parameter, however this is simply a hint
+to the `InputFormat` (which describes the input-specification for a Map-Reduce
+job), regarding the number of map tasks.
+
+Thus, e.g., if you expect 10TiB of input data and have 128MiB DFS blocks, you'll
+end up with 81920 maps, unless your `mapred.map.tasks` is even larger.
+Ultimately the `InputFormat` determines the number of maps.
+
+```
+Number of blocks = integer ceiling of input data over block size
+
+= ceil( input_data / block_size )
+= (10 * 2^40) / (2^7 * 2^20)
+= (10 * 2^40) / 2^27
+= 10 * 2^13
+= 10 * 8192
+= 81920
+```
+
+The question remains: What happens when a data record gets split into two
+different blocks?  After all, Hadoop has no idea about what the input data is
+going to look like, so it certainly can't distinguish where one record ends and
+the other begins.
+
+To deal with this, Hadoop uses something called _input splits_.  This is a
+logical representation of that data.  When a job client first calculates the
+input splits, it determines where the first complete record in a block begins,
+and where the last record ends.  In cases where the last record within a block
+has been split off, the input split includes location data on the remnants of
+the record, including the byte offset of the partial record.
+
+
+#### Point 3 & 4: Mapping & Handling Intermediate map Values
+
+For each map task, Hadoop places intermediate values into a circular buffer,
+located in memory, and is configurable under `mapreduce.task.io.sort.mb`
+(defaults to 100MB). This value is the total amount of data that a particular
+map task can output. If a configurable percentage of this limit is met, then the
+data will be spilled to the disk as a _shuffle spill file_.  Spills are written
+in a round-robin fashion.
+
+On any given node, multiple map tasks can be run, and as a result many spill
+files can be created.  Hadoop merges all of the spill files, sorts, and then
+partitions the resulting file based upon the number of reducers.  However, before
+an individual spill file is even written to disk, the spill thread does a few
+things:
+
+1. The spill file itself is partitioned based upon the number of reducers.
+2. Within each of the partitions of a spill file, the thread runs an in-memory
+sort by key.
+3. Finally, if a combiner function exists, it is run on the output of the
+resulting sort. The combiner allows for a more compact map output, leaving less
+data to write locally, thus less data to transfer to the reducer.  It is
+important to note that a combiner basically compresses the data, meaning it
+will eventually have to be decompressed before reducing.
```

Comment:
This makes perfect sense, but perhaps it would be good in the higher level overview to have mentioned that in addition to map() and reduce(), the user can declare a combine()?   Perhaps I am wrong though...